ecl/eclcc/WORKUNITS.rst

```diff
@@ -0,0 +1,862 @@
+context which of these is meant.
+
+Contents of a workunit
+======================
+
+A workunit is generated by the ecl compiler, and consists of a single dll/shared object.  That dll contains
+several different elements:
+
+* Various c++ helper classes, and exported factory methods are used to create instances of those classes.
+* An XML resource containing different pieces of information about a workunit, including workflow and graphs.
+* Other user-defined resources included in the manifest.
+
+The workunit dll contains everything that the engines need to execute the query.  When a workunit is executed, key
+elements of the xml information are cloned from the workunit dll and copied into a database (currently dali).  This
+is then extended with other information - e.g., input parameters, results, various statistics.  The IWorkUnit interface
+(define in common/workunit/workunit.hpp) is the public interface onto the contents, and hides many of the
+implementation details.
+
+How is it used?
+===============
+
+The workunit information is used by most of the components in the system.  The following is a quick outline:
+
+* eclcc
+  Creates a workunit dll from an ecl query.
+* eclccserver
+  Executes eclcc to create a workunit dll, and then clones some of the information into dali to create an active instance, ready to execute.
+* esp
+  Uses information in the workunit to publish workunits - including details of what parameters it takes, how they should be formatted, and what results it returns.
+* eclscheduler
+  Monitors workunits that are waiting for events, and updates them when those events occur.
+* eclagent/roxie
+  Process the different workflow actions
+* hthor/roxie/Thor
+  Execute graphs
+  
+Example
+*******
+
+The following ECL will be used as an example for the rest of the discussion.  It is a very simple search that 
+takes a string parameter 'searchName', which is the name of the person to search for, and returns the matching records
+from an index.  It also outputs the word 'Done!' as a separate result.
+
+::
+
+    STRING searchName := 'Smith' : STORED('searchName');
+    nameIndex := INDEX({ STRING40 name, STRING80 address }, 'names');
+    results := nameIndex(KEYED(name = searchName));
+    OUTPUT(results);
+    OUTPUT('Done!');
+
+Extracts from the XML and C++ that are generated from this example will be included in the following discussion.
+
+Workunit Main Elements
+**********************
+
+This section outlines the different sections in the workunit.  After that there is a walkthrough of the stages that
+take place when a workunit is executed, together with a more detailed explanation of the workunit contents.
+
+Workflow
+========
+The workflow is the highest level of control within a workunit.  It is used for two related purposes:
+
+- Scheduling
+  The HPCC system allows ECL code to be executed when certain events occur - e.g., every hour
```

Comment:
Hmm, looks like it depends if you are US or UK.  US does.  UK apparently doesn't.